cardano-cli/src/Cardano/CLI/EraBased/Options/Common.hs

```diff
@@ -2077,30 +2077,62 @@ pInvalidBefore = fmap SlotNo $ asum
     ]
   ]
 
-pInvalidHereafter :: Parser SlotNo
-pInvalidHereafter =
+pLegacyInvalidHereafter :: Parser SlotNo
+pLegacyInvalidHereafter =
   fmap SlotNo $ asum
-  [ Opt.option (bounded "SLOT") $ mconcat
-    [ Opt.long "invalid-hereafter"
-    , Opt.metavar "SLOT"
-    , Opt.help "Time that transaction is valid until (in slots)."
-    ]
-  , Opt.option (bounded "SLOT") $ mconcat
-    [ Opt.long "upper-bound"
-    , Opt.metavar "SLOT"
-    , Opt.help $ mconcat
-      [ "Time that transaction is valid until (in slots) "
-      , "(deprecated; use --invalid-hereafter instead)."
+    [ Opt.option (bounded "SLOT") $ mconcat
+      [ Opt.long "invalid-hereafter"
+      , Opt.metavar "SLOT"
+      , Opt.help "Time that transaction is valid until (in slots)."
+      ]
+    , Opt.option (bounded "SLOT") $ mconcat
+      [ Opt.long "upper-bound"
+      , Opt.metavar "SLOT"
+      , Opt.help $ mconcat
+        [ "Time that transaction is valid until (in slots) "
+        , "(deprecated; use --invalid-hereafter instead)."
+        ]
+      , Opt.internal
+      ]
+    , Opt.option (bounded "SLOT") $ mconcat
+      [ Opt.long "ttl"
+      , Opt.metavar "SLOT"
+      , Opt.help "Time to live (in slots) (deprecated; use --invalid-hereafter instead)."
+      , Opt.internal
       ]
-    , Opt.internal
-    ]
-  , Opt.option (bounded "SLOT") $ mconcat
-    [ Opt.long "ttl"
-    , Opt.metavar "SLOT"
-    , Opt.help "Time to live (in slots) (deprecated; use --invalid-hereafter instead)."
-    , Opt.internal
     ]
-  ]
+
+pInvalidHereafter :: ()
+  => CardanoEra era
+  -> Parser (TxValidityUpperBound era)
```

Comment:
In era-based commands we want to bring the era-sensitive types closer to the CLI.

This means we parse `TxValidityUpperBound era` instead of `Maybe SlotNo`.

This allows the parser to determine if the options are valid for the given era.

For example `pInvalidHereafter` will add these options only in a Shelley based eras and will be absent for the `Byron` era.

In the Byron era, `TxValidityNoUpperBound` always returned and there are no options for it.